modules/core/src/main/scala/io/janstenpickle/hotswapref/HotswapRef.scala

```diff
@@ -48,31 +45,53 @@ object HotswapRef {
+  def apply[F[_], R](initial: Resource[F, R])(implicit F: Concurrent[F]): Resource[F, HotswapRef[F, R]] = {
 
-    /* Provision a [[cats.effect.kernel.Ref]] to count the number of active requests for `R`, and do not release
-     * the resource until the number of open requests reaches 0. This effectively blocks the
-     * `swap` method until the previous resource is no longer in use anywhere.
-     */
-    def makeResource(r: Resource[F, R]): Resource[F, (Ref[F, Long], R)] =
-      for {
-        r0 <- r
-        ref <- Resource.make(Ref.of(0L))(references => Clock[F].sleep(100.millis).whileM_(references.get.map(_ != 0)))
-      } yield ref -> r0
+    /** Secure a resource by enriching it with a semaphore-based lock and a unique token and by modifying its finalizer.
+      *
+      * The lock is acquired in shared mode when the resource is accessed (permits concurrent access but prohibits
+      * finalization) and in exclusive mode when the resource is finalized (prohibits access).
+      *
+      * The token is used during access for consistent read of the holder reference.
+      */
+    def secure(res: Resource[F, R]): Resource[F, (R, Lock[F], Unique.Token)] = for {
+      lock <- Resource.eval(Lock[F])
+      r <- Resource
+        .makeFull((poll: Poll[F]) => poll(res.allocated)) { case (_, release) => lock.exclusive.surround(release) }
+        .map(_._1)
+      token <- Resource.eval(Unique[F].unique)
+    } yield (r, lock, token)
 
     for {
-      (hotswap, r) <- Hotswap(makeResource(initial))
-      sem <- Resource.eval(Semaphore(1))
-      ref <- Resource.eval(Ref.of(r))
+      (hotswap, securedR) <- Hotswap(secure(initial))
+      holder <- Resource.eval(Ref.of(securedR))
+      sem <- Resource.eval(Semaphore(1L))
     } yield new HotswapRef[F, R] {
       override def swap(next: Resource[F, R]): F[Unit] =
-        sem.permit.use(_ => hotswap.swap(makeResource(next).evalTap(ref.set))).void
+        sem.permit.surround(hotswap.swap(secure(next).evalTap(holder.set))).void
+
+      override val access: Resource[F, R] = {
+
+        /** Access to the resource is protected by a shared-mode lock. The holder reference is read at least twice:
+          * first, to retrieve its content, and then, after acquiring the lock, to check if the content hasn't changed
+          * since the first read. If the holder has been swapped, the lock is released and the new content is passed
+          * to the next step of the loop. Otherwise, it's used to build the resulting `Resource`.
+          */
+        val step: (R, Lock[F], Unique.Token) => F[Either[(R, Lock[F], Unique.Token), (R, F[Unit])]] =
+          (r, lock, token) =>
+            F.uncancelable { poll: Poll[F] =>
+              poll(lock.shared.allocated).flatMap { case (_, lockRelease) =>
+                holder.get.flatMap { case (r1, lock1, token1) =>
+                  if (token != token1) lockRelease.as(Left((r1, lock1, token1)))
```

Comment:
Not sure if you should use `=!=` here from the `Eq[Token]` instance.

Reply:
@vasilmkd We only need to check that the ref content has changed. Firstly, I wanted to use `eq`/`ne` (referential equality) on `R`, but that would require `R <: AnyRef`, so I introduced a token. It doesn't have an `Eq` instance and `!=` on it works like `ne`.

But now I think I could just use `ne` on `Tuple` since the holder ref contains a tuple which is AnyRef.

Reply:
Token does have an Eq instance. It's used in the CE laws.

Reply:
Hmm, I couldn't find one.

Reply:
It should work implicitly. There's a `Hash[Token]` instance, and `Hash extends Eq`.

Reply:
Right, I've seen it. I just forgot `Hash extends Eq`.

Reply:
Honestly, I just learned this fact today for the first time. 😆

Reply:
Does this code have any dowsides?
```scala
        val step: ((R, Lock[F])) => F[Either[(R, Lock[F]), (R, F[Unit])]] = { case tup @ (r, lock) =>
          F.uncancelable { poll: Poll[F] =>
            poll(lock.shared.allocated).flatMap { case (_, lockRelease) =>
              holder.get.flatMap { case tup1 @ (r1, lock1) =>
                if (tup ne tup1) lockRelease.as(Left((r1, lock1)))
                else F.pure(Right((r, lockRelease)))
              }
            }
          }
        }
```

Reply:
Honestly, I feel like this code is relying on implementation details that the reference returned by `Ref` would be identical, and `Ref` never explicitly makes that guarantee. I don't know what to tell you. 😕

Reply:
If you really want to avoid the `Unique.Token`, you can check reference equality on the `Lock` objects (I don't approve this either 😆).

Reply:
Ok, let's make it with Token and Eq.

Reply:
I was going to ask the same RE `Token` and `Eq`, but looking at the default implementation of `Eq` for `Token` it uses the underlying Java hashcode and equals: 

```scala
  def fromUniversalHashCode[A]: Hash[A] =
    new Hash[A] {
      def hash(x: A) = x.hashCode()
      def eqv(x: A, y: A) = x == y
    }
```

My question is, seeing as we don't give the user the opportunity to change the implementation of `Eq`, does the indirection for the compiler introduce any runtime overhead, condsidering we are optimising `access` for frequent calls? Also, maybe less importantly, given the motivation and implementation of `Token` do we think the implementation of `Eq` is likely to change?

Reply:
`Token` is already used in other libraries for these purposes and I can guarantee you that the implementation of both `Token` and `Eq` won't change. The design of Cats Effect `Token` takes these things into consideration.

Also, `Token` is a simple `final class Token()`, whose `==` method delegates to `eq`, and the C2 compiler is able to inline all of these `eqv` and `==` calls to `eq` very easily. There is no reason to worry.

Reply:
And `Token` was explicitly made to support reference equality only.

Reply:
Sorry, I missed this when I made my comment below, thanks for clearing that up. I'm happy to use the `Eq` instance here then.